Use `html2canvas` to assist thumbnail generation.

xinan opened the pull request:
* Use `html2canvas` and use the old thumbnail functions to do cropping and resizing. 
 - Able to capture html output (except cross-origin images)
 - Able to capture multiple turtles.
 - Able to capture turtle skins.


Note: I reverted the camera button because it turns out to be more complicated than I thought. There are several severe bugs that needs to be solved.

xinan commented:
I wonder why one of the Travis thread fails.

xinan commented:
@davidbau Could we sort out the test problem and get this merge first? Let's skip the camera button for this PR. I'm having another branch for that.

davidbau commented:
OK.  (The tests are still failing for me after we get rid of v10 etc.)

xinan commented:
:tada: It passed after re-running.

davidbau commented on content/src/thumbnail.js:
```diff
@@ -61,12 +78,13 @@ function getImageInfo(canvas) {
 
   // Find the longer edge and make it a square.
   var longerEdge;
+  var diff = Math.abs((imageWidth - imageHeight) / 2);
```
While you're thinking through these cases
- Do you think we should adjust things so that we never grow an image if there are fewer than 128 pixels?  To me it feels like we should shrink but never grow to make a thumbnail.
- Should there be special handling when there are no pixels at all on the canvas?  (e.g., turtle is hidden, there is no output)

xinan replied:
Yes it should never grow. I would change that. 

Previously if there is no output there will be no canvas, then it will not do the capturing. However, using `html2canvas`, there will always be a result canvas even if it is blank. I would try to figure out a good way to do this.

davidbau replied:
It is OK not to do anything special case when there is just an unmoved turtle: the icon can be just a turtle.

I am asking about the situation where even the turtle is not present and the screen is totally blank: it seems that a totally blank icon (and maybe even an "almost blank" icon) would be confusing to see and use within a grid of icons.

xinan replied:
Yeah I got what you mean. I was referring to that. 



Best regards, Liu Xinan

On Fri, Jul 17, 2015 at 10:03 AM, David Bau <notifications@github.com>
wrote:

>> @@ -61,12 +78,13 @@ function getImageInfo(canvas) {
>>  
>>    // Find the longer edge and make it a square.
>>    var longerEdge;
>> +  var diff = Math.abs((imageWidth - imageHeight) / 2);
> It is OK not to do anything special case when there is just an unmoved turtle: the icon can be just a turtle.
> I am asking about the situation where even the turtle is not present and the screen is totally blank: it seems that a totally blank icon (and maybe even an "almost blank" icon) would be confusing to see and use within a grid of icons.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/PencilCode/pencilcode/pull/155/files#r34857025

davidbau commented on package.json:
```diff
@@ -64,6 +64,7 @@
     "filetype": "./content/src/filetype.js",
     "gadget": "./content/src/gadget.js",
     "guide": "./content/src/guide.js",
+    "html2canvas": "./content/lib/html2canvas.js",
```
I think we discussed this before, but I don't remember.  Is it possible to bring in html2canvas via npm?

xinan replied:
It is possible but it doesn't work yet :stuck_out_tongue: I have filed an issue for them.

davidbau commented:
In this commit, what version of html2canvas is brought in?  Since we're not using `npm` to install it, did you add it to `bower.json` so that `grunt update` updates it?

davidbau commented on content/src/thumbnail.js:
```diff
@@ -1,20 +1,50 @@
+      if (innerBody.children[i].tagName.toLowerCase() === 'samp' && (
+          innerBody.children[i].className !== 'turtlefield' ||
+          innerBody.children[i].id == '_testpanel')) {
+        innerBody.children[i].style.display = 'none';
```
Here maybe have a list of `hiddenElements`, and then accumulate the list as well as the old value of `elt.style.display` when you set each `elt.style.display = 'none'`.

Then below in `onRendered`, you can go through the same list and restore all the old `elt.style.display` values.

davidbau commented:
This is very nice work.  I like how the test is solid green now.  I left a couple other comments.

xinan commented:
Okay updated with response to all the suggestions above :smile: 

* Fixed intermittent test error.
* Add `html2canvas` into `bower.json` and `grunt bowercopy`.
* Updated browser-side to not to generate blank thumbnails. 
* Updated server-side to remove the existing thumbnail if a save request doesn't come with a thumbnail.
* Use an array to remember all the old `style.display` settings of the nodes that we are going to hide, and restore them when we are done capturing thumbnails.
* Made it not to zoom in when the thumbnail is too small, so we will not get a blur/pixelized thumbnail.

davidbau commented:
Looks great.  Merging!
